src/cudadecoder/cuda-online-pipeline-dynamic-batcher.cc

```diff
@@ -44,10 +51,76 @@ CudaOnlinePipelineDynamicBatcher::~CudaOnlinePipelineDynamicBatcher() {
 
 void CudaOnlinePipelineDynamicBatcher::Push(
     CorrelationID corr_id, bool is_first_chunk, bool is_last_chunk,
-    const SubVector<BaseFloat> &wave_samples) {
-  std::lock_guard<std::mutex> lk(chunks_m_);
-  chunks_.push_back({corr_id, is_first_chunk, is_last_chunk, wave_samples});
-  n_chunks_not_done_.fetch_add(1);
+    const SubVector<BaseFloat> &wave_samples_subv) {
+  std::lock_guard<std::mutex> lk(next_batch_and_backlog_m_);
+
+  // Try to add the chunk is the batch right away
+  if (!TryAddChunkToNextBatchDeepCopy(corr_id, is_first_chunk, is_last_chunk,
+                                      wave_samples_subv)) {
+    // If we cannot add the chunk to the next batch,
+    // put it in the backlog
+    Vector<BaseFloat> wave_samples(wave_samples_subv);  // deep copy
+    backlog_.push_back(
+        {corr_id, is_first_chunk, is_last_chunk, std::move(wave_samples)});
+  }
+  n_chunks_not_done_.fetch_add(1, std::memory_order_release);
```

Comment:
If my understanding is correct, `n_chunks_not_done_` is not _Synchronized-With_ other atomic operations anywhere. If that's the case, `std::memory_order_relaxed` is more appropriate on every access to it. No real difference on the x64 architecture, as it has quite a strict memory ordering, but ARM and even more so POWER9 reorder accesses wildly. I think we support ARM.

If you put it alone by itself into a 64-byte structure aligned at a 64-byte boundary, it won't cause any false sharing, and will flush only a single cacheline. Since it's one per decoder, not much memory waste, but the accesses will be faster.

Reply:
I added std::memory_order_release to protect the backlog_ write from being reordered after. Is it not necessary?

Reply:
The more I'm thinking about this, the more confused I feel... I was probably wrong and you're right, the backlog item counter should use the Acquire-Release semantics. In any case, better safe than sorry :)

Reply:
I was certainly wrong. After reading the C++11 spec, it is indeed true that both the compiler and the HW are allowed to reorder across atomics with the relaxed memory semantics.

I attempted to see if the compile reordering happens, and did not find any on godbolt that would. https://godbolt.org/z/djcnx6zab -- uncommenting line 25 pessimizes both x64 and armv8 code from 2 wide to 4 narrow stores, 2 before the atomic increment and 2 after, as the code is written. Then I found this: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/n4455.html -- quite an interesting read encouraging the compiler writers to optimize atomics. 6 year past, it did not happen. But it's not only allowed, but kinda encouraged, too.